server/datastore/mysql/hosts_test.go

```diff
@@ -5975,3 +5978,215 @@ func testHostsLoadHostByOrbitNodeKey(t *testing.T, ds *Datastore) {
 	require.True(t, loadFleet.IsOsqueryEnrolled())
 	require.True(t, loadFleet.MDMInfo.IsPendingDEPFleetEnrollment())
 }
+
+func checkEncryptionKeyStatus(t *testing.T, ds *Datastore, hostID uint, expected *bool) {
+	ExecAdhocSQL(t, ds, func(tx sqlx.ExtContext) error {
+		var actual *bool
+
+		row := tx.QueryRowxContext(
+			context.Background(),
+			"SELECT decryptable FROM host_disk_encryption_keys WHERE host_id = ?",
+			hostID,
+		)
+
+		err := row.Scan(&actual)
+		require.NoError(t, err)
+		require.Equal(t, expected, actual)
+		return nil
+	})
+}
+
+func testHostsSetOrUpdateHostDisksEncryptionKey(t *testing.T, ds *Datastore) {
+	host, err := ds.NewHost(context.Background(), &fleet.Host{
+		DetailUpdatedAt: time.Now(),
+		LabelUpdatedAt:  time.Now(),
+		PolicyUpdatedAt: time.Now(),
+		SeenTime:        time.Now(),
+		NodeKey:         ptr.String("1"),
+		UUID:            "1",
+		OsqueryHostID:   ptr.String("1"),
+		Hostname:        "foo.local",
+		PrimaryIP:       "192.168.1.1",
+		PrimaryMac:      "30-65-EC-6F-C4-58",
+	})
+	require.NoError(t, err)
+	host2, err := ds.NewHost(context.Background(), &fleet.Host{
+		DetailUpdatedAt: time.Now(),
+		LabelUpdatedAt:  time.Now(),
+		PolicyUpdatedAt: time.Now(),
+		SeenTime:        time.Now(),
+		NodeKey:         ptr.String("2"),
+		UUID:            "2",
+		OsqueryHostID:   ptr.String("2"),
+		Hostname:        "foo.local2",
+		PrimaryIP:       "192.168.1.2",
+		PrimaryMac:      "30-65-EC-6F-C4-59",
+	})
+	require.NoError(t, err)
+
+	err = ds.SetOrUpdateHostDiskEncryptionKey(context.Background(), host.ID, "AAA")
+	require.NoError(t, err)
+
+	err = ds.SetOrUpdateHostDiskEncryptionKey(context.Background(), host2.ID, "BBB")
+	require.NoError(t, err)
+
+	checkEncryptionKey := func(hostID uint, expected string) {
+		ExecAdhocSQL(t, ds, func(tx sqlx.ExtContext) error {
+			var actual string
+
+			row := tx.QueryRowxContext(
+				context.Background(),
+				"SELECT base64_encrypted FROM host_disk_encryption_keys WHERE host_id = ?",
+				hostID,
+			)
+
+			err := row.Scan(&actual)
+			require.NoError(t, err)
+			require.Equal(t, expected, actual)
+			return nil
+		})
+	}
+
+	h, err := ds.Host(context.Background(), host.ID)
+	require.NoError(t, err)
+	checkEncryptionKey(h.ID, "AAA")
+
+	h, err = ds.Host(context.Background(), host2.ID)
+	require.NoError(t, err)
+	checkEncryptionKey(h.ID, "BBB")
+
+	err = ds.SetOrUpdateHostDiskEncryptionKey(context.Background(), host2.ID, "CCC")
+	require.NoError(t, err)
+
+	h, err = ds.Host(context.Background(), host2.ID)
+	require.NoError(t, err)
+	checkEncryptionKey(h.ID, "CCC")
+
+	// setting the encryption key to an existing value doesn't change its
+	// encryption status
```

Comment:
Looks like you don't test what the comment says? You don't set the key to an existing value AFAICT.

Reply:
Ah nevermind, you set its status here, then you set the key to the same value, and then you check that it didn't change the status.

Reply:
Nah that was just me stopping reading too early, seems fine to me!